url49584c1d41_Arrem_GGS_tgz-pJ8-com_gamezgalaxy_test_console_MainJ8/src/com/gamezgalaxy/GGS/util/logger/Logger.java

```diff
@@ -0,0 +1,179 @@
+	}
+	/**
+	 * Create a new instance of Logger
+	 */
+	public Logger() {
+		
+	}
+	
+	/**
+	 * Set the owner
+	 * @param obj
+	 */
+	public void setParent(Object obj) {
+		owner = obj;
+	}
+
+	/**
+	 * Start writing logs
+	 * @throws FileNotFoundException
+	 *                              If the file to log to cant be found
+	 */
+	public void Start(boolean changefilename) throws FileNotFoundException {
+		if (changefilename) {
+			try {
+				ChangeFilePath(filepath);
+			} catch (IOException e) {
+				e.printStackTrace();
+			}
+		}
+		Running = true;
+		runner = new Log();
+		runner.start();
+	}
+	
+	/**
+	 * Stop writing logs
+	 * This code will block until all logs have been written to the file
+	 * @throws InterruptedException
+	 */
+	public void Stop() throws InterruptedException {
+		Running = false;
+		runner.join();
+		out.close();
+	}
+	
+	/**
+	 * Add a log to the logger
+	 * @param message The message to add
+	 */
+	public void Log(String message) {
+		synchronized(queue) {
+			if (!Running)
+				return;
+			queue.add(message);
+		}
+	}
+
+	/**
+	 * Change the filepath of the log file
+	 * @param newpath The new filepath
+	 * @throws IOException
+	 */
+	public void ChangeFilePath(String filename) throws IOException {
+		this.filepath = filename;
+		if (out != null)
+			out.close();
+		FileUtils.CreateIfNotExist(filepath);
+		out = new PrintWriter(filepath);
+	}
+	
+	/**
+	 * Change the file that the log will be written to.
+	 * If the logger is already running, then that logger will stop writing
+	 * and write what's left in queuing in the new file.
+	 * @param directory
+	 * @param filename
+	 * @throws IOException
+	 */
+	public void ChangeFilePath(String directory, String filename) throws IOException {
+		this.filepath = directory + File.separator + filename;
+		if (out != null)
+			out.close();
+		FileUtils.CreateIfNotExist(directory, filename);
+		out = new PrintWriter(filepath);
+	}
+	
+	private class Log extends Thread {
+
+		@Override
+		public void run() {
+			Calendar cal = Calendar.getInstance();
+			Iterator<String> it = null;
+			while (Running) {
+				synchronized(queue) {
+					if (out == null) {
+						try {
+							Thread.sleep(10);
+						} catch (InterruptedException e) { }
+						continue;
+					}
+					it=queue.iterator();
+					if (it == null)
+						continue;
+					while (it.hasNext()) {
+						cal = Calendar.getInstance();
+						String message = (String)it.next();
+						String date = dateFormat.format(cal.getTime());
+						String finalmessage = "[" + date + "] " + message;
+						if (owner != null && owner instanceof LogInterface) {
+							if (message.contains("ERROR"))
+								((LogInterface)owner).onError(finalmessage);
+							else
+								((LogInterface)owner).onLog(finalmessage);
+						}
+						if (out != null)
+							out.println(finalmessage);
+					}
+					queue.clear();
+					out.flush();
+				}
+				try {
+					Thread.sleep(100);
+				} catch (InterruptedException e) { }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)